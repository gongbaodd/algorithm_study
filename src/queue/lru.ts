import { Queue } from "../queue/lib";

export class LRU<T> {
  private hashSet = new Set<T>();

  private queue: Queue<T> = new Queue();

  private length = 0;

  private get isQueueFull() {
    return this.queue.size() >= this.length;
  }

  private enqueue(data: T) {
    if (this.isQueueFull) {
      this.hashSet.delete(this.queue.top as T);
      this.queue.deQueue();
    }

    this.queue.enQueue(data);
    this.hashSet.add(data);
  }

  constructor(length = 0) {
    this.length = length;
  }

  accessData(data: T) {
    if (this.hashSet.has(data)) {
      this.queue.remove(data);
      this.queue.enQueue(data);

      return;
    }

    this.enqueue(data);
  }

  get cache() {
    return this.hashSet;
  }
}
