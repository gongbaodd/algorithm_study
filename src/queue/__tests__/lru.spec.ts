import { LRU } from "../lru";

describe("LRU", () => {
  describe("access serial is [1,2,5,1,6,7,6]", () => {
    it("should cache [7, 6, 1]", () => {
      const lru = new LRU<number>(3);

      lru.accessData(1);
      lru.accessData(2);
      lru.accessData(5);
      lru.accessData(1);
      lru.accessData(6);
      lru.accessData(7);
      lru.accessData(6);

      expect(lru.cache).toEqual(new Set([7, 6, 1]));
      expect(lru).toMatchSnapshot();
    });
  });
});
