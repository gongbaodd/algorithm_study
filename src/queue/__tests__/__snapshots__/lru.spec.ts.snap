// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LRU access serial is [1,2,5,1,6,7,6] should cache [7, 6, 1] 1`] = `
LRU {
  "hashSet": Set {
    1,
    6,
    7,
  },
  "length": 3,
  "queue": Queue {
    "pEnd": LNode {
      "data": 6,
      "next": null,
    },
    "pHead": LNode {
      "data": 1,
      "next": LNode {
        "data": 6,
        "next": LNode {
          "data": 7,
          "next": LNode {
            "data": 6,
            "next": null,
          },
        },
      },
    },
  },
}
`;
